TPGSup/devTPG300.db

```diff
@@ -55,18 +43,272 @@ record(mbbi, "$(P)UNITS")
     field(ZRST, "mbar")
     field(ONST, "Torr")
     field(TWST, "Pa")
+    field(FLNK, "$(P)UNITS")
     field(SIML, "$(P)SIM")
     field(SIOL, "$(P)SIM:UNITS:SP")
     field(SDIS, "$(P)DISABLE")
 }
 
-alias("$(P)UNITS", "$(P)UNITS:SP:RBV")
+record(mbbo, "$(P)FUNCTION")
+{
+    field(DESC, "Switching function selection")
+    field(SCAN, "Passive")
+    field(VAL, "0")
+    field(ZRST, "1")
+    field(ONST, "2")
+    field(TWST, "3")
+    field(THST, "4")
+    field(FRST, "A")
+    field(FVST, "B")
+    field(FLNK, "$(P)FUNCTION:RB:FAN")
+}
+
+record(fanout, "$(P)FUNCTION:RB:FAN")
+{
+    field(DESC, "Fanout for all function readbacks")
+    field(SCAN, "Passive")
+    field(LNK1, "$(P)FUNCTION:SEL:RB")
+    field(LNK2, "$(P)FUNCTION:1:RB")
+    field(LNK3, "$(P)FUNCTION:2:RB")
+    field(LNK4, "$(P)FUNCTION:3:RB")
+    field(LNK5, "$(P)FUNCTION:4:RB")
+    field(LNK6, "$(P)FUNCTION:A:RB")
+    field(LNK7, "$(P)FUNCTION:B:RB")
+}
+
+record(stringin, "$(P)FUNCTION:SEL:RB")
+{
+    field(DESC, "Selected function readback")
+    field(DTYP, "stream")
+    field(SCAN, "Passive")
+    field(PINI, "YES")
+    field(INP, "@devTPG300.proto getFunction($(P)FUNCTION.VAL) $(PORT)")
+    field(FLNK, "$(P)FUNCTION:SEL:RB:FAN")
+    field(SIML, "$(P)SIM")
+    field(SIOL, "$(P)SIM:FUNCTION:SEL:RB")
+    field(SDIS, "$(P)DISABLE")
+}
+
+record(ai, "$(P)FUNCTION:LOW:SP")
+{
+    field(DESC, "Setpoint for function's low threshold")
+    field(SCAN, "Passive")
+    field(PREC, "1")
+    field(LOPR, "0")
+    field(HOPR, "9.9")
+    field(FLNK, "$(P)FUNCTION:LOW:SP:CALC")
+}
+
+record(calc, "$(P)FUNCTION:LOW:SP:CALC")
+{
+    field(SCAN, "Passive")
+    field(PREC, "1")
+    field(INPA, "$(P)FUNCTION:LOW:SP")
+    field(INPB, "$(P)FUNCTION:LOW:SP.LOPR")
+    field(INPC, "$(P)FUNCTION:LOW:SP.HOPR")
+    field(INPD, "$(P)FUNCTION:LOW:SP.PREC")
+    field(CALC, "FLOOR(D*10*MIN(MAX(A,B),C))/(D*10)")
+}
+
+record(longin, "$(P)FUNCTION:LOW:E:SP")
+{
+    field(DESC, "Setpoint for function's low threshold exponent")
+    field(SCAN, "Passive")
+    field(LOPR, "-99")
+    field(HOPR, "99")
+    field(FLNK, "$(P)FUNCTION:LOW:E:SP:CALC")
+}
+
+record(calc, "$(P)FUNCTION:LOW:E:SP:CALC")
+{
+    field(SCAN, "Passive")
+    field(INPA, "$(P)FUNCTION:LOW:E:SP")
+    field(INPB, "$(P)FUNCTION:LOW:E:SP.LOPR")
+    field(INPC, "$(P)FUNCTION:LOW:E:SP.HOPR")
+    field(CALC, "MIN(MAX(A,B),C)")
+}
+
+record(ai, "$(P)FUNCTION:HIGH:SP")
+{
+    field(DESC, "Setpoint for function's high threshold")
+    field(SCAN, "Passive")
+    field(PREC, "1")
+    field(LOPR, "0")
+    field(HOPR, "9.9")
+    field(FLNK, "$(P)FUNCTION:HIGH:SP:CALC")
+}
+
+record(calc, "$(P)FUNCTION:HIGH:SP:CALC")
+{
+    field(SCAN, "Passive")
+    field(PREC, "1")
+    field(INPA, "$(P)FUNCTION:HIGH:SP")
+    field(INPB, "$(P)FUNCTION:HIGH:SP.LOPR")
+    field(INPC, "$(P)FUNCTION:HIGH:SP.HOPR")
+    field(INPD, "$(P)FUNCTION:HIGH:SP.PREC")
+    field(CALC, "FLOOR(D*10*MIN(MAX(A,B),C))/(D*10)")
```

Comment:
What is the justification for needing this? If the device expects a certain precision we can just set that in streamdevice?